In [1]:
import pandas as pd
import numpy as np
from pathlib import Path


In [2]:
import os
print(os.getcwd())

c:\Users\HP\Desktop\projects\New_ara\coffee-flavor-map


In [3]:
file_path = Path(r"C:\Users\HP\Desktop\projects\New_ara\coffee-flavor-map\data\raw\simplified_coffee.csv")

df = pd.read_csv(file_path)

# Raw Data Discovery — Coffee Reviews

**Goal:** Understand the structure, quality, and content of the corpus before cleaning (Step 1.2).  
You will write the code in each section below.

## 1. Shape and columns

**Discover:** How many rows and columns? What are the column names and dtypes?

**What to check:** Enough rows for a small Word2Vec corpus (100+); a column that holds the review text (e.g. `review`).

In [7]:
# Your code: df.shape, df.columns, df.dtypes (or df.info())

In [8]:
df.shape

(1246, 9)

In [9]:
df.columns

Index(['name', 'roaster', 'roast', 'loc_country', 'origin', '100g_USD',
       'rating', 'review_date', 'review'],
      dtype='object')

In [10]:
df.info

<bound method DataFrame.info of                                   name                     roaster  \
0             Ethiopia Shakiso Mormora                Revel Coffee   
1                   Ethiopia Suke Quto                 Roast House   
2           Ethiopia Gedeb Halo Beriti   Big Creek Coffee Roasters   
3              Ethiopia Kayon Mountain  Red Rooster Coffee Roaster   
4     Ethiopia Gelgelu Natural Organic   Willoughby's Coffee & Tea   
...                                ...                         ...   
1241                    Finca Patzibir                El Gran Cafe   
1242                   Proyecto Aurora                El Gran Cafe   
1243                  Finca El Potrero                El Gran Cafe   
1244          Chacayá Santiago Atitlán                El Gran Cafe   
1245                    Espresso No. 3                El Gran Cafe   

             roast    loc_country     origin  100g_USD  rating    review_date  \
0     Medium-Light  United States   Ethiopia  

## 2. The review column (your corpus source)

**Discover:** The column that contains flavor text. How many non-null values? Any missing/empty reviews?

**What to check:** Count of nulls; maybe drop nulls and see how many rows remain. Sample 2–3 raw review strings to see length and format.

In [11]:
# Your code: null count for review column; optionally df['review'].dropna(); sample rows with .sample() or .iloc
df.isnull().sum()

name            0
roaster         0
roast          12
loc_country     0
origin          0
100g_USD        0
rating          0
review_date     0
review          0
dtype: int64

In [12]:
df.sample(frac=0.25)

,name,roaster,roast,loc_country,origin,100g_USD,rating,review_date,review
282,Colombia Pink Bourbon,Paradise Roasters,Medium-Light,United States,Colombia,7.33,94,June 2019,"Elegant, tropical, rich-toned. Strawberry guav..."
992,Ethiopia Guji Washed G1 Uncle Select Lot 21/01,GK Coffee,Medium-Light,Taiwan,Ethiopia,8.80,93,January 2022,"Crisply sweet, floral-toned. Jasmine, cocoa ni..."
178,La Union (Colombia),Barrington Coffee Roasting,Medium-Light,United States,Colombia,4.54,91,October 2018,"Crisply sweet, chocolaty. Baker’s chocolate, p..."
229,Panama Ninety Plus Perci Lot 50,Bird Rock Coffee Roasters,Medium-Light,United States,Panama,26.46,97,January 2019,"Wild, unruly, utterly brilliant and original. ..."
657,Guatemala Acetenango Gesha Lot 2,RamsHead Coffee Roasters,Medium-Light,United States,Guatemala,8.80,93,January 2021,"Cocoa-and spice-toned. Cocoa powder, cherry, f..."
...,...,...,...,...,...,...,...,...,...
358,Ethiopia Konga Natural,Blues Brew Coffee,Medium-Light,Taiwan,Ethiopia,6.40,93,November 2019,"Richly sweet berry-toned. Ripe strawberry, lil..."
147,Ethiopia Kayon Mountain,Big Creek Coffee Roasters,Medium-Light,United States,Ethiopia,4.70,93,October 2018,"Intricate, richly sweet. Lemon verbena, raspbe..."
914,Colombia La Riviera Sudan Rume,RamsHead Coffee Roasters,Medium-Light,United States,Colombia,10.57,94,November 2021,"Delicate, complex, rich-toned. Cocoa nib, pers..."
975,Ka’u Bourbon Champagne Natural,Paradise Roasters,Medium-Light,United States,Hawai'I,26.46,94,January 2022,"Tropical, floral, and fruit-toned. Star jasmin..."


In [13]:
df.iloc[0]

name                                    Ethiopia Shakiso Mormora
roaster                                             Revel Coffee
roast                                               Medium-Light
loc_country                                        United States
origin                                                  Ethiopia
100g_USD                                                     4.7
rating                                                        92
review_date                                        November 2017
review         Crisply sweet, cocoa-toned. Lemon blossom, roa...
Name: 0, dtype: object

## 3. Text length and variety

**Discover:** How long are the reviews (e.g. word count or character count per row)? Min, max, median.

**What to check:** Whether you have short snippets vs long paragraphs; this will inform how you split into sentences later (Step 1.2).

In [15]:
# Your code: e.g. series of word counts per review (split on spaces), then .describe() or .min()/.max()/.median()
df.describe()

,100g_USD,rating
count,1246.000000,1246.000000
mean,10.479125,93.306581
std,13.486674,1.479413
min,0.170000,84.000000
25%,5.260000,93.000000
50%,6.170000,93.000000
75%,9.600000,94.000000
max,132.280000,97.000000


In [ ]:
df['review'].str.split().str.len()

0       46
1       36
2       48
3       40
4       45
        ..
1241    35
1242    32
1243    42
1244    34
1245    39
Name: review, Length: 1246, dtype: int64

In [21]:
df['review'].str.split().str.len().max()

np.int64(72)

In [22]:
df['review'].str.split().str.len().min()

np.int64(7)

In [18]:
df['review'].str.split().str.len().mean()

np.float64(37.630016051364365)

## 4. Metadata (optional but useful later)

**Discover:** Value counts for categorical columns: `roast`, `origin`, `loc_country`. Rating range.

**What to check:** Balance of roast levels and origins; you may filter or stratify by these when building the corpus or evaluating.

In [23]:
# Your code: df['roast'].value_counts(), df['origin'].value_counts(), df['rating'].min()/max()
df['roast'].value_counts()

roast
Medium-Light    904
Light           198
Medium          114
Medium-Dark      15
Dark              3
Name: count, dtype: int64

In [24]:
df['origin'].value_counts()


origin
Ethiopia                            447
Colombia                            151
Kenya                               143
Hawai'I                             113
Guatemala                            96
Panama                               70
Costa Rica                           46
Indonesia                            28
Brazil                               15
El Salvador                          14
Democratic Republic Of The Congo     14
Yemen                                13
Peru                                 13
Rwanda                               12
Burundi                              12
Mexico                               11
Nicaragua                            11
Taiwan                                9
Ecuador                               8
Honduras                              6
Nepal                                 2
Bolivia                               2
Dominican Republic                    2
Thailand                              2
Uganda                           

## 5. Noise and encoding

**Discover:** Any obvious noise in the text? Special characters, extra spaces, or encoding issues (e.g. mojibake)?

**What to check:** Print a few full review strings (e.g. `df['review'].iloc[0]`). Look for `\n`, `  `, or weird symbols. Decide what you’ll normalize in the cleaning step.

In [31]:
for i in range(3): 
    print(f"Review {i}: {repr(df['review'].iloc[i])}")


Review 0: 'Crisply sweet, cocoa-toned. Lemon blossom, roasted cacao nib, date, rice candy, white peppercorn in aroma and cup. Savory-tart structure; delicate, silky mouthfeel. The richly drying finish leads with cocoa-toned, crisply sweet floral notes in the short and a hint of spice (white peppercorn) in the long.'
Review 1: 'Delicate, sweetly spice-toned. Pink peppercorn, date, myrrh, lavender, roasted cacao nib in aroma and cup. Crisp, spice-toned structure with citrus-like acidity; satiny, very smooth mouthfeel. The crisply sweet finish centers around spice and pungent floral notes.'
Review 2: 'Deeply sweet, subtly pungent. Honey, pear, tangerine zest, dark chocolate, pistachio in aroma and cup. Sweet and juicy yet crisp in structure (think ripe pear); plush and buoyant in mouthfeel. Resonant and flavor saturated in the short finish; simplifies around hints of tangerine and pistachio in the long.'


## 6. Summary

**Conclusion (from your discovery results):**

- **Usable review texts:** 1,246 — no nulls in the `review` column. (12 nulls in `roast` only; no impact on the corpus.)

- **Structure:** One full paragraph per row (not one sentence). Each review is ~7–72 words (median ~38): short-to-medium tasting notes (aroma/cup, structure, mouthfeel, finish). Good for Word2Vec as-is; for sentence-level analysis you may split on punctuation (e.g. `.` or `;`) in the cleaning step.

- **Cleaning step (next):** Text looks clean (English, no mojibake in the sample). Plan: (1) normalize encoding when reading/writing if needed; (2) optionally split each review into sentences for more training units; (3) strip leading/trailing whitespace; (4) write one segment per line to `data/processed/corpus.txt`. Metadata (roast, origin) is available for later filtering or stratification.

In [ ]:
# Optional: one number — total word count across all reviews (for corpus size estimate)

# Task 2 — Clean Text and Build Corpus (Step 1.2)

**Goal:** Turn the raw `review` column into a clean corpus file at `data/processed/corpus.txt` (one segment per line), ready for Word2Vec in Step 1.3.  
You will implement each step below.

## 2.1 Normalize and strip

**Task:** From `df['review']`, produce a series/list of strings where each string has: (a) leading/trailing whitespace removed, (b) no empty strings kept. Optionally normalize encoding (e.g. ensure UTF-8) if you saw issues in Task 1.

**What to check:** After this step you have only non-empty, trimmed strings. Drop or filter out any row where the review becomes empty after strip.

In [33]:
# Your code: e.g. .str.strip(), dropna or filter empty, optionally .str.encode('utf-8', errors='ignore')
df_reviews = df["review"].dropna()

In [34]:
import string
df_reviews = (df['review']
                      .str.lower()
                      .str.replace(f'[{string.punctuation}]', '', regex=True)
                      .str.strip())

In [35]:
df_reviews

0       crisply sweet cocoatoned lemon blossom roasted...
1       delicate sweetly spicetoned pink peppercorn da...
2       deeply sweet subtly pungent honey pear tangeri...
3       delicate richly and sweetly tart dried hibiscu...
4       hightoned floral dried apricot magnolia almond...
                              ...                        
1241    crisply sweet nuttoned almond brittle pie cher...
1242    chocolaty floralframed cocoa nib honeysuckle d...
1243    hightoned enticingly sweet black cherry narcis...
1244    vibrantly sweet subtly nuanced apricot dark ch...
1245    evaluated as espresso multilayered complex dar...
Name: review, Length: 1246, dtype: object

## 2.2 (Optional) Split into sentences

**Task:** Decide: use each review as one line (paragraph-level), or split on sentence boundaries (e.g. `. ` or `; `) so each line is one sentence. If you split, flatten the result so you have a list of sentences, not a list of lists.

**What to check:** For Word2Vec, more shorter lines often help. Use a simple rule (e.g. split on `. `) or a library (e.g. nltk.sent_tokenize) if you prefer. Output: a single list (or generator) of text segments.

In [38]:
# Your code: either keep one paragraph per line, or split each review into sentences and flatten to one list of strings
# Your code: Path or os.makedirs for data/processed; open(..., 'w', encoding='utf-8'); write one segment per line
sentences = df['review'].str.split(r'(?<=[.!?])\s+').explode().tolist()

# Clean up any empty strings or purely whitespace entries
sentences = [s.strip() for s in sentences if len(s.strip()) > 0]
sentences

['Crisply sweet, cocoa-toned.',
 'Lemon blossom, roasted cacao nib, date, rice candy, white peppercorn in aroma and cup.',
 'Savory-tart structure; delicate, silky mouthfeel.',
 'The richly drying finish leads with cocoa-toned, crisply sweet floral notes in the short and a hint of spice (white peppercorn) in the long.',
 'Delicate, sweetly spice-toned.',
 'Pink peppercorn, date, myrrh, lavender, roasted cacao nib in aroma and cup.',
 'Crisp, spice-toned structure with citrus-like acidity; satiny, very smooth mouthfeel.',
 'The crisply sweet finish centers around spice and pungent floral notes.',
 'Deeply sweet, subtly pungent.',
 'Honey, pear, tangerine zest, dark chocolate, pistachio in aroma and cup.',
 'Sweet and juicy yet crisp in structure (think ripe pear); plush and buoyant in mouthfeel.',
 'Resonant and flavor saturated in the short finish; simplifies around hints of tangerine and pistachio in the long.',
 'Delicate, richly and sweetly tart.',
 'Dried hibiscus, fine musk, almon

## 2.3 Write corpus to disk

**Task:** Write your list of text segments to `data/processed/corpus.txt`, one segment per line. Ensure the `data/processed/` directory exists (create it if needed). Use UTF-8 encoding when writing.

**What to check:** After running, the file exists; opening it shows one review (or one sentence) per line; no extra blank lines unless you intend them.

In [41]:
processed_path = Path(r"C:\Users\HP\Desktop\projects\New_ara\coffee-flavor-map\data\processed")
processed_path.mkdir(parents=True, exist_ok=True)
file_path = processed_path / 'corpus.txt'

with open(file_path, 'w', encoding='utf-8') as f:
    for s in sentences:
        f.write(f"{s}\n")

print(f"Successfully saved {len(sentences)} segments to {file_path}")


Successfully saved 5093 segments to C:\Users\HP\Desktop\projects\New_ara\coffee-flavor-map\data\processed\corpus.txt


## 2.4 Verify output

**Task:** Load the first few lines of `data/processed/corpus.txt` and print them. Count total lines (number of segments). Optionally count total words in the file.

**What to check:** Format matches what you expect; line count is reasonable (e.g. 1246 if one review per line, or more if you split into sentences).

In [42]:
# Your code: open corpus.txt, read first N lines, print; count lines (e.g. with a loop or sum(1 for line in file))
with open(file_path, 'r', encoding='utf-8') as f:
    for _ in range(5):
        print(f.readline().strip())


Crisply sweet, cocoa-toned.
Lemon blossom, roasted cacao nib, date, rice candy, white peppercorn in aroma and cup.
Savory-tart structure; delicate, silky mouthfeel.
The richly drying finish leads with cocoa-toned, crisply sweet floral notes in the short and a hint of spice (white peppercorn) in the long.
Delicate, sweetly spice-toned.


## 2.5 Task 2 summary

**Conclusion (from your Task 2 results):**

- **Lines (segments) in `corpus.txt`:** 5,093 — one sentence per line. You split each review on sentence boundaries (regex on `.` `!` `?` followed by whitespace), then stripped and dropped empty strings, so 1,246 paragraphs became 5,093 segments.

- **Paragraph vs sentence:** You split into sentences (Option 2.2). That gives Word2Vec more, shorter training units and better co-occurrence signal for flavor terms.

- **Issues:** None observed — UTF-8 write/read is fine; sample lines look clean. Next step is **Step 1.3**: train Word2Vec on this corpus (e.g. one sentence = one document in the iterator).

In [ ]:
# Optional: total word count in corpus.txt (for Step 1.3 reference)